### ISSUE TITLE
API for Multi-Step loaders

### ISSUE THREAD

**Issue by kylebarron** [visgl/loaders.gl#1140]
Many file formats are well suited to a multi step process. First, read metadata about the resource, then use that metadata in further requests.

Possible loaders (existing or potential) that could fit into this approach:

- File archives, such as `ZipLoader`: First pass loads ZIP metadata and the central directory; second pass loads one or more files from within the ZIP archive.
- Vector files with multiple layers, such as a zipped Shapefile: First pass parses available layers, second pass loads a single layer's data
- Tiled vector datasets, such as a `TileJSONLoader`: First pass loads the TileJSON document itself, second pass loads individual tiles, delegating to the `ImageLoader` or `MVTLoader` as necessary (a TileJSON doc can represent either images or vector tiles).
- Tiled raster datasets, such as GeoTIFF or [Zarr](https://zarr.readthedocs.io/en/stable/): First pass loads metadata, second pass handles tile requests and georeferencing the tile using the initial metadata

The crucial difference between existing loaders API and a multi-step loading API is that some amount of state needs to be saved between the initial request and following requests.

One idea might be a function like `loadMetadata`, which would return metadata that could be passed back to the loader function.

```js
// Fetch just ZIP metadata and store that in an object
const zipLoaderMetadata = loadMetadata(zipUrl, ZIPLoader);

// Pass that metadata back to the ZIPLoader again, and pass option data derived from metadata
const fileData = load(zipLoaderMetadata, ZIPLoader, {zip: {filename: 'foo.txt'}});
```

Loaders that support this might have a `loadMetadata` key on the loader definition. If `load` is called directly on such loaders, `load` could first call `loadMetadata` internally to then pass on to `load`.

**Comment by ibgreen** [comment]
I feel there might be two cases here:
1. the metadata is an individual but related resource - like TileJSON metadata file for tiles 
2, the metadata is extracted from the single file being loaded, at which time the application needs to make further decisions about exactly what to load.

In the first state, we are looking for a generalized way to help the app manage state between loads, which is now done adhoc by various tile layer helper classes.

In the second case, the loader may need a mechanism for asking the app for more information after a load has started. Unless it is acceptable for the resource is to be loaded twice (first partially for header information).

One could imagine a batched loader that calls a callback after reading the header batch, confirming which parts of the file the app actually wants to be streamed.

**Comment by kylebarron** [comment]
I don't see that much of a dichotomy. To me, whether the metadata is within the same file or in a different file, it still has to be parsed before proceeding. The only issue is that for single files, you either make multiple requests to a single resource, or if you already have the entire data in memory, that's stored while waiting for the second step.

For example in the ZIP file case, if the loader receives a URL to a remote file, I'd expect to first make a partial read or two to retrieve the table of contents, plus additional reads for more data. That means loaders.gl would be able to load items from a ZIP archive, no matter the full size of the ZIP archive.

> One could imagine a batched loader that calls a callback after reading the header batch, confirming which parts of the file the app actually wants to be streamed.

This assumes that you know what's in the file before you open it. If you need to check what files/layers exist, you really need two steps.

**Comment by ibgreen** [comment]
> To me, whether the metadata is within the same file or in a different file, it still has to be parsed before proceeding. 

Well, the difference is that loaders.gl currently has no support for multi-file loading. We do not have an abstraction that lets you treat a vector tileset as a single entity (e.g. providing an interface like the ZipLoader). Perhaps creating loader abstractions that let users treat a multi-file datasource as a single thing using some common conventions is what you want to propose. It may be a good idea but a different set of work from extracting header data from single files.

> This assumes that you know what's in the file before you open it. If you need to check what files/layers exist, you really need two steps.

Depend on the format. If header information comes early on it can be done in a single pass. For zip files the header information comes at the end. We can still stream batches if the source is random access like a `File` object, but not for a general `Response` obv.

**Comment by kylebarron** [comment]
> multi-file loading

I think that's part of what I'm getting at, but in my view as long as the server supports HTTP range requests, multiple files and multiple pages within a file is practically equivalent. E.g. ZIP and GeoTIFFs are "internally tiled", and TileJSON and Zarr are "externally tiled" but the abstraction should (or at least could) be the same.

> If header information comes early on it can be done in a single pass

But not if the user needs to look at what files exist? Imagine a file format like ZIP but metadata came at the beginning, and the file is 400MB. Once the user opens the file, they realize they only need layers stored at the end of the file. Then 300MB of the file might be downloaded unnecessarily, and two separate range requests would be much faster. But two passes would be needed if you can't proceed until you know what layers/files exist.

> currently has no support for multi-file loading

It does in the `tile-3d` layers right? It's just not generalized?

**Comment by ibgreen** [comment]
> But not if the user needs to look at what files exist? Imagine a file format like ZIP but metadata came at the beginning, and the file is 400MB. Once the user opens the file, they realize they only need layers stored at the end of the file. Then 300MB of the file might be downloaded unnecessarily, and two separate range requests would be much faster. But two passes would be needed if you can't proceed until you know what layers/files exist.

Sure for very large file formats it would be possible to create random access file loaders that could take advantage of range requests. Another layer of complexity on the loaders but certainly possible. (I actually have a wip branch on a random access ZipLoader, that starts loading chunks from the end and goes from there...)

> It does in the tile-3d layers right? It's just not generalized?

True though that code is so specific to the 3D tile formats that I consider it part of the application layer, completely outside of the loaders architecture. Almost like a convenience class that happens to be packaged in loaders.gl. FWIW, my current hope was to unify the Tile2D and Tile3D traversal code, perhaps still a reasonable first step?

**Comment by kylebarron** [comment]
> Another layer of complexity on the loaders but certainly possible

You're right it would be more complexity. At this point I just want to discuss the overall API, and in the future we can make per-loader decisions whether the increased complexity is worth it.

> unify the Tile2D and Tile3D traversal code, perhaps still a reasonable first step

Last I heard from Xiaoji I think she was worried about bundle size implications because the Tile3d traversal had more code. I think that's more complex than a first step needs to be, and it's also not clear to me that traversal code should live in loaders.gl.

I mentioned a `TileJSONLoader` in the first comment, but I think for simplicity, such a loader should receive tile indices in the second step, so that traversal lives outside. E.g. something like

```js
const tileJsonMetadata = await loadMetadata(tileJsonUrl, TileJSONLoader);
const tileData = await load(tileJsonMetadata, TileJSONLoader, {tileJson: {x: 100, y: 200, z: 10}});
```

**Comment by ibgreen** [comment]
> Last I heard from Xiaoji I think she was worried about bundle size implications

There are ways around that. 

> And it's also not clear to me that traversal code should live in loaders.gl.

The goal of loaders.gl is to be a framework independent solution for loading key formats. Traversal is a huge part of he effort of implementing I3S and 3D Tiles. Removing that support would make those loaders rather low value to non-deck.gl use cases.

> such a loader should receive tile indices in the second step, so that traversal lives outside. E.g. something like

Most use cases are not a two step loading, they need to reload tiles again and again as viewports update. Basically a generalized traversal mechanism like the one in the tile layers. Caching is needed. If the tileJSONMetadata value is the complete Tileset class sure this could work.

**Comment by kylebarron** [comment]
I was thinking about this a little more today and one possibility that doesn't require any change in the global loaders API is to have _two_ loaders for some formats. So a zarr loader might have both `ZarrMetadataLoader` and `ZarrLoader`. `ZarrLoader` will call `ZarrMetadataLoader` under the hood of some options key doesn't exist (maybe `options.zarr.loaderMetadata`). This means that basic users don't need to care that the `MetadataLoader` exists and can use `ZarrLoader` normally, while advanced users can call `load(ZarrMetadataLoader)` directly and pass its output to multiple calls of `load(ZarrLoader, {options: {zarr: {loaderMetadata}}})`. Decoupling these calls means that each additional call to `load(ZarrLoader)` doesn't need to load the original metadata file.

So the `ZarrLoader` high-level implementation might look like the following:

```js
function loadZarr(url, options) {
  let loaderMetadata = options?.zarr?.loaderMetadata;
  if (!loaderMetadata) {
    loaderMetadata = load(url, ZarrMetadataLoader);
  }
  ...
}
```

This side steps for now the issue of traversal, which could take a while to discuss in general, and unblocks progress on other loaders that _could_ be enhanced with a multi-step process: GeoTIFF, ZIP, Zarr, GeoPackage

**Comment by kylebarron** [comment]
@zakjan I'd love your feedback on this 😄

**Comment by zakjan** [comment]
Thanks for pinging me. I'm interested in this because of a possibility to have a native support for rendering tiled data sources (COG) with TileLayer.renderSublayers.

I think it's important that multi-step loading is designed as an optional enhancement of a simpler single-step loading, so that simpler single-step can be implemented first for a new data source, and user can easily switch between them when both are available.

Possibly TileLayer URL template logic could be extracted to a loader as well, so that there is a single interface for providing tiles to TileLayer?

Support for multi-dimensional data seems also interesting. From UI perspective, I can see these usecases:
- show a list of dimensions (layers, bands) - this seems could be supported by calling XxxMetadataLoader separately
- select a dimension - XxxLoader would need to accept a dimension (layer, band) id to load

**Comment by kylebarron** [comment]
> I think it's important that multi-step loading is designed as an optional enhancement of a simpler single-step loading, so that simpler single-step can be implemented first for a new data source, and later, when both are available, a developer can easily switch between them.

💯 

> URL template logic could be extracted to a loader as well

I don't think there's much URL template logic in deck.gl? It's just replacing `{x}`, `{y}` and `{z}`.

The tile visibility/loading logic, however, is more complex, and it's possible we could move that here to loaders, since we already have that here for the 3d tiles. Ib has some thoughts about unifying the 2d and 3d tile logic as well.

**Comment by kylebarron** [comment]
> Any comments on the approach above? [#1140 (comment)](https://github.com/visgl/loaders.gl/issues/1140#issuecomment-855042176)

So for a COG it could be something like

```js
const metadata = await load(geotiffUrl, GeoTiffMetadataLoader)

const tile = load(null, GeoTiffLoader, {geotiff: {metadata, tile: [0, 0, 0]}}
```

**Comment by zakjan** [comment]
> I don't think there's much URL template logic in deck.gl? It's just replacing {x}, {y} and {z}.
> The tile visibility/loading logic, however, is more complex, and it's possible we could move that here to loaders, since we already have that here for the 3d tiles. Ib has some thoughts about unifying the 2d and 3d tile logic as well.

Yes, I think that the entire logic for tile visibility calculation and tile loading might be worth extracting, because the URL template represents a single tiled data source.

Re COG loading and rendering, do you think that it could be used in a similar way, by providing URL only to TileLayer?

```js
new TileLayer({
  data: 'cog.tif',
  renderSubLayers: props => {
    const {bbox: {west, south, east, north}} = props.tile;
    return new BitmapLayer(props, {
      data: null,
      image: props.data,
      bounds: [west, south, east, north],
    });
  },
});
```

**Comment by kylebarron** [comment]
> do you think that it could be used in a similar way, by providing URL only to TileLayer?

I think we'd want a separate issue to really get into the possibility of rendering COGs like so, since this issue is more just for API design. There's so much complexity underlying your code snippet there that I don't think will ever go into the `TileLayer` directly. More likely a `COGTileLayer` in https://github.com/visgl/deck.gl-community.

In particular, here you ideally want the shape of the tiles to be driven by the COG, not top-down from the TileLayer. Otherwise you'd have to at least piece together multiple tiles from the COG (and likely do reprojection) to create each Web Mercator tile for the TileLayer.

**Comment by zakjan** [comment]
I understand that the current TileLayer is tailored for XYZ URL tiles. I haven't explored the tiling code in detail yet, but that's exactly what I thought might be worth extracting, so that TileLayer could be repurposed (or a new layer created) to be able to work with any tiled source. Similarly to how COG loader would contain COG tiling logic, XYZ URL loader would contain XYZ URL tiling logic. It would follow the separation of concerns between deck.gl and loaders.gl. Depends if it is realistic to achieve, if tiling can be abstracted in such way.

Otherwise, there is going to need to be a loader-specific XxxTileLayer for each tiled data source.

**Comment by kylebarron** [comment]
Yes, it makes sense to incubate such a tile layer within deck.gl-community. I think the main key is you'd need a common denominator to describe an arbitrary tiling structure from any data source. Something we could consider using is the [OGC 2D TileMatrixSet standard](https://www.ogc.org/standards/tms), which defines a 2D hierarchy of tiles in any projection. Then each tiled loader (COG, Zarr, TileJSON) could create such a TMS object to describe the source data layout and then a `TMSTileLayer` could compute tile visibility based on the provided TMS document.

**Comment by ibgreen** [comment]
The new `Source` and `Service` architecture in 4.0 handles the multi-step loader use case. There are examples in WMS and PMtiles. More documentation will be forthcoming as the design stabilizes.
